File: Q2611_「プラス・マイナス・ゼロ」問題/Answer.hs
-- Try CodeIQ
-- author: Leonardone @ NEETSDKASU

import qualified Data.IntMap as IntMap

main = print . solve . read =<< getLine where
    solve n = loop n $ IntMap.fromList [(0,1)]
    loop 0 xm = IntMap.findWithDefault 0 0 xm
    loop i xm = loop (i - 1) nx where
        nx = IntMap.foldrWithKey f IntMap.empty xm
        f k v rm = res where
            tmp = IntMap.insertWith (+) (k + i) v rm
            res = IntMap.insertWith (+) (k - i) v tmp

File: Q2611_「プラス・マイナス・ゼロ」問題/Answer.clj
; Try CodeIQ
; author: Leonardone @ NEETSDKASU
; ============================================
(loop [n (read) a (assoc {} 0 1)]
      (if (zero? n) (println (get a 0 0))
          (let [h (reduce #(let [k (key %2) 
                                 v (val %2)
                                 k1 (+ k n)
                                 k2 (- k n)
                                 v1 (+ v (get %1 k1 0))
                                 v2 (+ v (get %1 k2 0))]
                             (assoc %1 k1 v1 k2 v2)) {} a)]
          (recur (dec n) h))))

